chemprop/cli/hyperopt.py

```diff
@@ -1,16 +1,356 @@
+import json
+import logging
+import sys
 from argparse import ArgumentParser, Namespace
+from pathlib import Path
 
+import torch
+from lightning import pytorch as pl
+from ray import tune
+from ray.train import CheckpointConfig, RunConfig, ScalingConfig
+from ray.train.lightning import (
+    RayDDPStrategy,
+    RayLightningEnvironment,
+    RayTrainReportCallback,
+    prepare_trainer,
+)
+from ray.train.torch import TorchTrainer
+from ray.tune.schedulers import ASHAScheduler
+
+try:
+    from ray.tune.search.hyperopt import HyperOptSearch
+except ImportError:
+    NO_HYPEROPT = True
+
+try:
+    from ray.tune.search.optuna import OptunaSearch
+except ImportError:
+    NO_OPTUNA = True
+
+from chemprop.cli.common import add_common_args, process_common_args, validate_common_args
+from chemprop.cli.train import (
+    add_train_args,
+    build_datasets,
+    build_model,
+    build_splits,
+    normalize_inputs,
+    process_train_args,
+    validate_train_args,
+)
 from chemprop.cli.utils.command import Subcommand
+from chemprop.data import MolGraphDataLoader
+from chemprop.nn import AggregationRegistry
+from chemprop.nn.utils import Activation
+
+
+logger = logging.getLogger(__name__)
 
+DEFAULT_SEARCH_SPACE = {
+    "activation": tune.choice(categories=list(Activation.keys())),
+    "aggregation": tune.choice(categories=list(AggregationRegistry.keys())),
+    "aggregation_norm": tune.quniform(lower=1, upper=200, q=1),
+    "batch_size": tune.loguniform(lower=16, upper=256, base=2),
+    "depth": tune.randint(lower=2, upper=6),
+    "dropout": tune.choice([tune.choice([0.0]), tune.quniform(lower=0.05, upper=0.4, q=0.05)]),
+    "ffn_hidden_size": tune.quniform(lower=300, upper=2400, q=100),
+    "ffn_num_layers": tune.randint(lower=2, upper=6),
```

Comment:
Is there a reason we need to explore such large MLP predictors? If an MPNN is flexible enough to effectively learn representation, then surely we don't need to project these representation into some very high dimensional latent space multiple times prior to making a prediction, i.e., I should be able to learn a representation that is optimal w/r/t an affine transformation or 1-layer MLP.   I worry that we're implicitly pushing users towards overparameterized models based on random hyperoptimization splits

Reply:
Chas did a pretty throrough study on this: https://github.com/chemprop/chemprop/pull/571
I set the hyperparameter search range based on his PR

Reply:
Those hyperopt studies:
1. were done using IID splitting. This generally favors overparameterized models as noise in the train distribution can more easily be fit and used at test-time. This is not desirable but empirically favored for these types of splits.
2. show no statiscally significant difference for 4/5 datasets and only marginally improved results for the 5th

My point is we should think about the implications: theoretical, empirical, and budgetary, of setting our default search space to be such large models. They become significantly larger than the current default model:

| layer size | # layers | # params | notes |
| --------- | -------- | --------- | ------ |
| 300 | 1 | 319k | **default** |
| 300 | 2 | 770k | **smallest possible** |
| 1200 | 4 | 4.9M | **middle of the road** |
| 2400 | 6 | 39M | **largest possible** |

I seriously doubt that typical users (whom we should be targeting) can honestly benefit from a model with 40M parameters on their datasets.